### ISSUE TITLE
"BIN file was corrupted, please download it again" should smartly resume and perhaps not reload entire game

### ISSUE THREAD

**Issue by ell1e** [sharkwouter/minigalaxy#630]
Thanks to the awesome #615 I got to try to download a few multi part games again, but I ran into another curious corner case: I got a "BIN file was corrupted, please download it again" error for Fallout 3 GOTY. This seems a bit strange since even with an extremely shaky connection, TLS should usually ensure integrity. But assuming that this really was caused by a connection hiccup anyway, rather than some bug in the downloader, e.g. the downloader not transparently doing a HTTPS resume when the download fell short, there seem to be two things that are a bit odd about what happens after:

1. The message appearing in the first place. Usually, for a smart downloader e.g. like Steam, the established behavior would be to simply quietly resume redownloading the affected part. The user interaction doesn't really seem to be required here.

2. The way the recovery is done when redownloading. When I prompt it to restart, minigalaxy seems to then proceed to start over the entire game from scratch, rather than just the affected BIN file, at least from how the progress bar looks like. Especially for bigger games, which are now not uncommon on GOG, this seems like it doesn't really scale well if that's how errors are handled.

**Affected minigalaxy version:** 1.3.1 (via Flathub)

**Comment by GB609** [comment]
The progress bar is a matter on its own.
But i just found a pretty serious bug in the download_manager related to file resume while reworking the code.
See https://github.com/sharkwouter/minigalaxy/pull/655#issuecomment-2676735067

The way the download_manager checks if a download stream has to be consumed effectively checks:
`downloaded_part < request.header('content-length')`

When resuming a partial download from disc, the to download will include a startpoint.
Naturally this means that the response will NOT include any bytes up until the start point, so the returned content length will also be smaller by that amount.
This ultimately means that any download that has progressed up to exactly half will not be finished when restarting minigalaxy in-between. The response stream is just never consumed in that case.

**Comment by ell1e** [comment]
There seems to be also at least one or possibly more hidden conditions where the resume fails, rather than completing as corrupted.

**Comment by GB609** [comment]
Maybe. But it's hard to debug if there's multiple reasons at once.

**Comment by GB609** [comment]
> 
>     1. The message appearing in the first place. Usually, for a smart downloader e.g. like Steam, the established behavior would be to simply quietly resume redownloading the affected part. The user interaction doesn't really seem to be required here.
> 
> **Affected minigalaxy version:** 1.3.1 (via Flathub)

The problem here is the point in time when corruption is checked currently.
It happens when the installer is started, not after the download procedure.

In a way, this makes sense. 
File corruption during download is one thing. But in case the disc or file system has issues, an already downloaded file can also become corrupted later. However, when the file will be checked on usege/installation anyway, checking a freshly downloaded file for corruption directly seems kind of redundant.
Also, both functions are more or less independent of each other right now. If minigalaxy would also check the download, we'd either have to more strongly couple download + install to tell the installer that the checksums has been checked already, or endure double checking for all downloaded files, which can take quite a while, depending on the type of disc and cpu of the machine.

**Comment by ell1e** [comment]
The problem is that if you let the installer do it, minigalaxy doesn't know which part is affected. This is a problem for redownloading just the affected bit.

**Comment by GB609** [comment]
No, the installer actually knows the checksums. They are requested from the api on demand. The only gap right now is when fully offline afaik.

**Comment by ell1e** [comment]
I perhaps found another bug today, when I left my computer unattended for a while downloading a game, today it just wasn't downloading anymore. There is no indication what happened, the progress bar is just gone and the UI looks like I never began installing the game. It seems like perhaps there was some interruption and the resume didn't work as intended.

**Comment by GB609** [comment]
This is a very vague description you are providing here.
The most probable cause is that there was some connection issue (or your disk running out of space). The retry here is a bit basic (and also doesn't wait after failure).
This normally leads minigalaxy to cancel the download after a while and (currently) delete the downloaded file.
Or there could have been an error during installation.

So, if possible, please provide a bit more info:
1. Which game
2. What _exactly_ was the UI for that game like? Did the button say "download" or "install" or did it maybe even show that there was an update?
3. is there a non-empty directory for this game in your `GOG Games/` folder for it?
4. Was it a windows or linux game
5. Is there any partial file download in `XDG_CACHE_HOME`/minigalaxy/download/<gamename>`?

**Comment by ell1e** [comment]
Sorry for providing too little info!
 
1. The game was Fallout 4! The disk wasn't full. An internet interruption happening previously however is quite likely (my internet is a bit shaky).
3. It says "Download", no update icon, it looks exactly like all the other games I haven't ever attempted to download and install. It's like I never had begun downloading it. Last time I saw it, the download had been progressed to 40% or something.
4.  No
5. Windows game!
6. Apparently, yes:    ```
    /.var/app/io.github.sharkwouter.Minigalaxy/cache/minigalaxy$ ls download/Fallout\ 4\ Game\ of\ the\ Year\ Edition/
   Fallout4GameoftheYearEdition-9.bin
   setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-6.bin
   setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-7.bin
   setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-8.bin
   setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-9.bin
   ```

**Comment by GB609** [comment]
Thank you for the input.

Given how many files there are, at least some of them got downloaded, at least partially. Then one ran into an error which canceled the entire list of downloads.

Just for reference, could you also provide a listing of the file sizes and a screenshot from the gog library download section for the offline installer of this game? I'd like to compare and see if any of the files above were completed.

Guess i gotta work a bit on error handling for the download_manager in general.

**Comment by ell1e** [comment]
I hope this is helpful:
```
~/.var/app/io.github.sharkwouter.Minigalaxy/cache/minigalaxy/download/Fallout 4 Game of the Year Edition$ ls
Fallout4GameoftheYearEdition-9.bin
setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-6.bin
setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-7.bin
setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-8.bin
setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-9.bin
~/.var/app/io.github.sharkwouter.Minigalaxy/cache/minigalaxy/download/Fallout 4 Game of the Year Edition$ stat 'setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-6.bin'
  File: setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-6.bin
  Size: 4294967294	Blocks: 8388608    IO Block: 4096   regular file
Device: 23h/35d	Inode: 15714776    Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/   ellie)   Gid: ( 1000/   ellie)
Access: 2025-03-01 06:58:25.572310513 +0100
Modify: 2025-03-01 10:14:33.993388946 +0100
Change: 2025-03-01 10:14:33.993388946 +0100
~/.var/app/io.github.sharkwouter.Minigalaxy/cache/minigalaxy/download/Fallout 4 Game of the Year Edition$ stat 'setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-7.bin'
  File: setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-7.bin
  Size: 4294967294	Blocks: 8388608    IO Block: 4096   regular file
Device: 23h/35d	Inode: 15714777    Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/   ellie)   Gid: ( 1000/   ellie)
Access: 2025-03-01 02:59:52.908237267 +0100
Modify: 2025-03-01 10:50:41.624677074 +0100
Change: 2025-03-01 10:50:41.624677074 +0100
~/.var/app/io.github.sharkwouter.Minigalaxy/cache/minigalaxy/download/Fallout 4 Game of the Year Edition$ stat 'setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-8.bin'
  File: setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-8.bin
  Size: 4294967294	Blocks: 8388608    IO Block: 4096   regular file
Device: 23h/35d	Inode: 15714775    Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/   ellie)   Gid: ( 1000/   ellie)
Access: 2025-03-01 02:59:48.465858892 +0100
Modify: 2025-03-01 08:48:44.170298207 +0100
Change: 2025-03-01 08:48:44.170298207 +0100
~/.var/app/io.github.sharkwouter.Minigalaxy/cache/minigalaxy/download/Fallout 4 Game of the Year Edition$ stat 'setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-9.bin'
  File: setup_fallout_4_game_of_the_year_edition_1.10.163.0_(64bit)_(66700)-9.bin
  Size: 224688813 	Blocks: 438848     IO Block: 4096   regular file
Device: 23h/35d	Inode: 15714778    Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/   ellie)   Gid: ( 1000/   ellie)
Access: 2025-02-28 12:29:35.216390848 +0100
Modify: 2025-02-28 13:54:21.676780063 +0100
Change: 2025-02-28 13:54:21.676780063 +0100
```

![Image](https://github.com/user-attachments/assets/34ec7f3c-0d0d-4216-a5bf-23977fb78074)

**Comment by ell1e** [comment]
It seems like when I press "Download" it actually resumes 🤔 despite in between looking like it wouldn't.

**Comment by ell1e** [comment]
Ignore this if this additional small incident isn't helpful: so okay, it got stuck "In Queue" all of a sudden some hours later, and I restarted minigalaxy, and then it actually gave me an error message that the install couldn't be done due to lacking disk space (sadly I didn't write down the exact wording). Except there is enough disk space, and it was only 55%ish into the download, so that doesn't make much sense. Clicking "Download" after that though seems to still resume with the partial 55% it previously had, so it seems like it kept the partial files at least which is super nice!

**Comment by GB609** [comment]
When a download starts, files (partial or full otherwise) are only deleted when they differ from the url resource. This is done with a pretty rudimentary check for equality of a small block at the beginning of the file. Not optimal, but if we'd do a full check of every byte, it would be the same as starting from the beginning again. Checksums would be pointless on partial files as we dont get such checksums from remote. Maybe another look at this is required.

The fix you quoted is one of those to repair some long-standing bugs with download corruption that appears when resuming a download after restart. Nothing more. The code that decides if a file has to be deleted runs before the fixed code is reached.
It's possible that there are more errors, but thats something we need to test and investigate.

**Comment by GB609** [comment]
> Okay, this time it stopped again with the UI reverted as if it were never installed or partially downloaded, and when I pressed "Download" again multiple gigabytes worth of partial setup files seem to have been deleted. (I'm not quite sure if they were deleted when it stopped, or at the point I pressed "Download" again, but I checked right after that and they were gone.) It left the complete 4GB chunks though, it seems only the incomplete ones which were a bunch of around 2GB each, were deleted. **So under some circumstances it definitely seems to delete huge parts of what it already downloaded,** which seemingly makes it potentially impossible to ever finish the download if that happens.
> 
> Edit: this was with 1.3.2 (latest flatpak stable as of today)
> 
> Edit 2: was this perhaps https://github.com/sharkwouter/minigalaxy/pull/656/files this problem? Since most of the affected chunks seem to have been beyond 50% 🤔

@ell1e 
After going over the code again, i am not entirely sure how the error you describe happens.
As long as you don't manually click cancel, mg should normally resume a download as long as it is not complete.
Partial files are only deleted when there is a download error (although it think this should not happen on error automatically) that does not go away even after multiple retries (the retries are made directly one after each other currently, which is a bug. There should be a bit of sleep or wait in-between).
But it could also be related to the fact that you're using the flatpak version. I don't know how the flatpak sets up the cache dir. Maybe there's some quota on that, or it's placed as in-memory. Or maybe it's not persisting stuff correctly. Which means that files will potentially be gone after restart of mg. 

I might be able to say more with a detailed debug log.
For that, you would have to run
`flatpak -env=MG_DEBUG=1 <your-minigalaxy-flatpak-command>` to get debug output onto the console. You could also combine this with tee and/or redirection to pipe into a file because it will be a lot of output.
But please don't openly post the log file here without a bit of post-editing. It's likely that your gog token is included in some lines when errors appear.

**Comment by ell1e** [comment]
> Partial files are only deleted when there is a download error (although it think this should not happen on error automatically) that does not go away even after multiple retries

That seems like the culprit, and generally like a bad idea, unless I'm misunderstanding. Can you explain why that would be useful to do unless the user explicitly cancels the downloading? Wouldn't this result if e.g. an overnight download runs into a one hour internet outage, that everything is deleted? I think this should be removed. Steam doesn't do this either, it'll happily sit through hours of outages and simply resume once internet is back up.

**Comment by GB609** [comment]
> > Partial files are only deleted when there is a download error (although it think this should not happen on error automatically) that does not go away even after multiple retries
> 
> That seems like the culprit, and generally like a bad idea, unless I'm misunderstanding. Can you explain why that would be useful to do unless the user explicitly cancels the downloading? Wouldn't this mean if e.g. an overnight download runs into a one hour internet outage, that everything is deleted? I think this should be removed. Steam doesn't do this either, it'll happily sit through hours of outages and simply resumes once internet is back up.
> 
> Edit: I'm 99% sure the affected flatpak dirs have no quota and no automatic cleaning.

No, not everything is deleted. Only the files that were downloading in that moment. 

I certainly agree that this is a bad idea either way and i'm currently in the progress of changing all that, including a proper UI to give more control. I assume it was made like that because the code didn't really differentiate between manual cancel and failure, so stuff is deleted to prevent the cache from filling up the disk with user-canceled partial downloads (if the game is not downloaded/isntalled at a later time).

**Comment by ell1e** [comment]
> Only the files that were downloading in that moment.

That does match what I was seeing, all the partial files were gone, the full 4GB ones remained. This can still easily reset a user back enough that they never manage to complete a file, depending on how their connection is like. So I think that this is likely the cause.

**Comment by ell1e** [comment]
I wonder, did the resume fix ship already? If not then ignore this message, but if it did, I'm still seeing corrupted downloads sometimes after resume with latest 1.3.2 on flathub. I wonder what the cause could be. This time the connection wasn't even really shaky, I just quit minigalaxy at some point and later reopened for a resume, and it seems like that broke something.

**Comment by GB609** [comment]
I fixed one very big issue i just stumbled across in 1.3.2, but there might other issues we just haven't found yet.

I know there is some issue with the option 'keep installers' and dlc downloads.

Downloads might also become corrupted when the game is very large and gets an update in-between. So a part of the files is from the old version and other files from the new version. Integrity checks done when installing could then fail because the old files might not match the new checksums.

To fix this, we'd have to save a bit more info about ongoing downloads as we do right now.
